File: ios/Pylux/Services/PsCloudOwnership.swift
// SPDX-License-Identifier: LicenseRef-AGPL-3.0-only-OpenSSL

import Foundation

/// Raw entitlement fields from Sony internal_entitlements API.
struct PsCloudEntitlement {
    let id: String
    let productId: String
    let activeFlag: Bool
    let packageType: String
    let name: String
    let conceptId: String
    let featureType: Int   // PSN feature_type: 3=full game, 1=trial/free, 0=add-on/DLC
}

enum PsCloudOwnership {
    static let pageSize = 300
    static let pageCooldownSeconds: TimeInterval = 0.1

    private struct CatalogIndex {
        var byProductId: [String: Int] = [:]
        var byConceptId: [String: Int] = [:]
    }

    static func filterOwnedPs5Games(_ entitlements: [PsCloudEntitlement]) -> [PsCloudEntitlement] {
        entitlements.filter { ent in
            // Previously required packageType == "PSGD" (PS5 only), which dropped owned
            // PS4 titles (e.g. God of War 2018) and PS3 titles. Accept every active game
            // entitlement; streamability is enforced downstream by the catalog cross-reference
            // (matches are deduped by conceptId), so non-streamable / add-on entitlements are
            // harmlessly dropped there.
            guard ent.activeFlag else { return false }
            let pid = ent.productId
            guard !pid.hasPrefix("IP"), !pid.hasPrefix("SUB") else { return false }
            // Hide EXTRAS: feature_type==0 is DLC / add-ons / themes / avatars / cross-buy "tracks",
            // never a base game (games are feature_type 1=trial/free or 3/5=full). Safe: can't hide a
            // game. Trials/free and full games are kept; the trial-vs-full split is handled at merge.
            guard ent.featureType != 0 else { return false }
            return true
        }
    }

    static func crossReferenceOwnedGames(
        filteredEntitlements: [PsCloudEntitlement],
        publicCatalog: [CloudGame],
        plusLibrarySupplement: [CloudGame] = [],
        productIdAliases: [String: String] = [:],
        componentIdsByProductId: [String: [String]] = [:]
    ) -> [CloudGame] {
        var catalogMap: [String: CloudGame] = [:]
        for game in publicCatalog {
            catalogMap[game.id] = game
        }
        for (alias, canonical) in productIdAliases {
            if catalogMap[alias] != nil { continue }
            if let meta = catalogMap[canonical] {
                catalogMap[alias] = meta
            }
        }
        var supplementMap: [String: CloudGame] = [:]
        for game in plusLibrarySupplement {
            supplementMap[game.id] = game
        }

        let browseStableKey = buildStableKeyIndex(publicCatalog)
        let supplementStableKey = buildStableKeyIndex(plusLibrarySupplement)
        let browseByConcept = buildConceptIdIndex(publicCatalog)
        let supplementByConcept = buildConceptIdIndex(plusLibrarySupplement)

        var byKey: [String: CloudGame] = [:]
        var byKeyRank: [String: Int] = [:]

        // Enrich one matched catalog row into an owned CloudGame and dedupe it into byKey, keeping
        // OUR convention (conceptId+PLATFORM dedupe, canonical-entitlement rank). Called once for a
        // direct match, or once per component for a bundle (upstream PR #15 bundle-sibling matching).
        func emit(_ meta: CloudGame, _ ent: PsCloudEntitlement) {
            let displayName = meta.name.isEmpty ? ent.name : meta.name
            let game = CloudGame(
                productId: meta.id,
                name: displayName,
                imageUrl: meta.imageUrl,
                landscapeImageUrl: meta.landscapeImageUrl,
                platform: meta.platform,
                serviceType: meta.serviceType,
                conceptUrl: meta.conceptUrl,
                conceptId: meta.conceptId,
                isOwned: true,
                entitlementId: ent.id,
                storeProductId: ent.productId,
                featureType: ent.featureType
            )
            let key = ownedDedupeKey(meta: meta, ent: ent)
            let candidateRank = ownedStreamRank(ent)
            if byKey[key] != nil {
                // Keep the best streaming candidate: the canonical full-game entitlement (its
                // product_id is the real streamable game, not a DLC/bonus product Gaikai rejects).
                if candidateRank > (byKeyRank[key] ?? -1) {
                    byKey[key] = game
                    byKeyRank[key] = candidateRank
                }
            } else {
                byKey[key] = game
                byKeyRank[key] = candidateRank
            }
        }

        for ent in filteredEntitlements {
            let stable = productIdStableKey(ent.productId)
            let entStable = productIdStableKey(ent.id)
            let skipStableDemo = ent.name.localizedCaseInsensitiveContains("demo")
            let meta: CloudGame?
            if !ent.productId.isEmpty, let g = catalogMap[ent.productId] {
                meta = g
            } else if !ent.id.isEmpty, let g = catalogMap[ent.id] {
                meta = g
            } else if !ent.conceptId.isEmpty, let g = browseByConcept[ent.conceptId] {
                // conceptId is region-stable; product IDs are region-prefixed (EP9000 vs UP9000).
                meta = g
            } else if !ent.conceptId.isEmpty, let g = supplementByConcept[ent.conceptId] {
                meta = g
            } else if !ent.productId.isEmpty, ent.id == ent.productId,
                      let g = supplementMap[ent.productId] {
                meta = g
            } else if let stable, !skipStableDemo, let g = browseStableKey[stable] {
                meta = g
            } else if let stable, !skipStableDemo, let g = supplementStableKey[stable] {
                meta = g
            } else if let entStable, !skipStableDemo, let g = browseStableKey[entStable] {
                // Stable-key match on the ENTITLEMENT id (upstream PR #15): catches cross-gen / upgrade
                // entitlement ids whose stable key matches a catalog row even when product_id did not.
                meta = g
            } else if let entStable, !skipStableDemo, let g = supplementStableKey[entStable] {
                meta = g
            } else {
                meta = nil
            }

            if let meta {
                emit(meta, ent)
                continue
            }

            // Bundle-sibling expansion (upstream PR #15): a bundle entitlement (e.g. RE7 Gold) has no
            // direct catalog row, but its component entitlement ids each map to a component game.
            var seenPids = Set<String>()
            for siblingId in componentIdsByProductId[ent.productId] ?? [] {
                let siblingMeta: CloudGame?
                if let g = catalogMap[siblingId] {
                    siblingMeta = g
                } else if let g = supplementMap[siblingId] {
                    siblingMeta = g
                } else if let sStable = productIdStableKey(siblingId), !skipStableDemo {
                    siblingMeta = browseStableKey[sStable] ?? supplementStableKey[sStable]
                } else {
                    siblingMeta = nil
                }
                guard let sMeta = siblingMeta, !sMeta.id.isEmpty, !seenPids.contains(sMeta.id) else { continue }
                seenPids.insert(sMeta.id)
                emit(sMeta, ent)
            }
        }

        return Array(byKey.values)
    }

    // Edition identity = conceptId + PLATFORM (matching the catalog's edition key), so a cross-gen
    // title owned on both PS4 and PS5 stays as two separate library entries instead of collapsing
    // into one. Same-platform duplicate SKUs (a remaster's add-ons) still merge.
    private static func ownedDedupeKey(meta: CloudGame, ent: PsCloudEntitlement) -> String {
        if !meta.conceptId.isEmpty { return "c:\(meta.conceptId):\(platformToken(ent.productId))" }
        if !meta.id.isEmpty { return "p:\(meta.id)" }
        if !ent.id.isEmpty { return "e:\(ent.id)" }
        return "u:\(meta.id):\(ent.id)"
    }

    // Platform token from a product id (CUSA = PS4, PPSA = PS5).
    static func platformToken(_ productId: String) -> String {
        if productId.contains("PPSA") { return "ps5" }
        if productId.contains("CUSA") { return "ps4" }
        return ""
    }

    // A "full game" entitlement (vs add-on/avatar/theme): PSN marks the base game with a *GD
    // package_type (PSGD/PS4GD); add-ons use PS4MISC/PSAL/etc.
    private static func isFullGameEntitlement(_ ent: PsCloudEntitlement) -> Bool {
        ent.featureType == 3 || ent.packageType.hasSuffix("GD")
    }

    // Rank an owned entitlement as THE streaming candidate for its edition (higher = preferred).
    // Bonus/upgrade SKUs collapse to the same conceptId+platform as the base game; package/feature
    // flags don't disambiguate (Death Stranding DC's "Bonus Content" is also PSGD + feature_type 3).
    // The reliable signal: the base game's entitlement id EQUALS its product_id, while bonus/upgrade
    // SKUs carry a different id -- so prefer the canonical full-game entitlement.
    private static func ownedStreamRank(_ ent: PsCloudEntitlement) -> Int {
        var rank = 0
        if !ent.productId.isEmpty && ent.productId == ent.id { rank += 4 } // canonical base-game SKU
        if isFullGameEntitlement(ent) { rank += 2 }
        if !ent.id.isEmpty { rank += 1 }
        return rank
    }

    // conceptId + platform for an owned/catalog game; the owned product id (storeProductId) takes
    // precedence so the owned edition's platform is used, else the catalog product id.
    private static func conceptPlatformKey(_ game: CloudGame) -> String {
        guard !game.conceptId.isEmpty else { return "" }
        let pid = game.storeProductId.isEmpty ? game.id : game.storeProductId
        return "\(game.conceptId)|\(platformToken(pid))"
    }

    /// Tokenize on '-' and '_'; identity is all tokens except the last (store SKU).
    private static func productIdStableKey(_ productId: String) -> String? {
        guard !productId.isEmpty else { return nil }
        var tokens: [String] = []
        for dashPart in productId.split(separator: "-") {
            for token in dashPart.split(separator: "_") where !token.isEmpty {
                tokens.append(String(token))
            }
        }
        guard tokens.count >= 2 else { return nil }
        return tokens.dropLast().joined(separator: "|")
    }

    private static func buildStableKeyIndex(_ games: [CloudGame]) -> [String: CloudGame] {
        var index: [String: CloudGame] = [:]
        for game in games {
            guard let key = productIdStableKey(game.id) else { continue }
            if index[key] == nil {
                index[key] = game
            }
        }
        return index
    }

    private static func buildConceptIdIndex(_ games: [CloudGame]) -> [String: CloudGame] {
        var index: [String: CloudGame] = [:]
        for game in games where !game.conceptId.isEmpty {
            if index[game.conceptId] == nil {
                index[game.conceptId] = game
            }
        }
        return index
    }

    /// Normalize a conceptId (imagic encodes it as a number) to a non-empty string, else nil.
    static func conceptIdString(_ value: Any?) -> String? {
        if let i = value as? Int { return i > 0 ? String(i) : nil }
        if let d = value as? Double { return d > 0 ? String(Int(d)) : nil }
        if let s = value as? String, !s.isEmpty { return s }
        return nil
    }

    static func mergeOwnedIntoBrowseCatalog(
        browseCatalog: [CloudGame],
        ownedCrossRef: [CloudGame],
        addUnmatched: Bool = true   // false = only mark ownership on catalog entries (Catalog tab)
    ) -> [CloudGame] {
        var games = browseCatalog
        var catalogIndex = buildCatalogIndex(games)

        for owned in ownedCrossRef {
            // Trials / free-to-play (feature_type 1) are kept as their OWN card so the user can Stream
            // the trial/free build, while the full version still shows separately as a not-owned
            // "Add Game" card -- so a trial must NOT collapse into the full-game catalog entry.
            let isTrialTier = owned.featureType == 1
            let catalogMatch = isTrialTier ? -1 : findCatalogIndexForOwned(owned, catalogIndex: catalogIndex)
            if catalogMatch >= 0 {
                var existing = games[catalogMatch]
                existing.isOwned = true
                if !owned.entitlementId.isEmpty { existing.entitlementId = owned.entitlementId }
                if !owned.storeProductId.isEmpty { existing.storeProductId = owned.storeProductId }
                games[catalogMatch] = existing
                continue
            }

            guard addUnmatched else { continue }
            var entry = owned
            entry.isOwned = true
            registerInCatalogIndex(entry, index: games.count, catalogIndex: &catalogIndex)
            games.append(entry)
        }

        return games.sorted {
            if $0.isOwned != $1.isOwned { return $0.isOwned && !$1.isOwned }
            return $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending
        }
    }

    static func parseEntitlement(_ obj: [String: Any]) -> PsCloudEntitlement? {
        guard let id = obj["id"] as? String, !id.isEmpty else { return nil }
        let gameMeta = obj["game_meta"] as? [String: Any] ?? [:]
        let name = (gameMeta["name"] as? String) ?? id
        let conceptId = conceptIdString(gameMeta["conceptId"])
            ?? conceptIdString(gameMeta["concept_id"])
            ?? conceptIdString(obj["conceptId"])
            ?? ""
        return PsCloudEntitlement(
            id: id,
            productId: (obj["product_id"] as? String) ?? "",
            activeFlag: (obj["active_flag"] as? Bool) ?? false,
            packageType: (gameMeta["package_type"] as? String) ?? "",
            name: name,
            conceptId: conceptId,
            featureType: (obj["feature_type"] as? NSNumber)?.intValue ?? 0
        )
    }

    private static func buildCatalogIndex(_ games: [CloudGame]) -> CatalogIndex {
        var catalogIndex = CatalogIndex()
        for i in games.indices {
            registerInCatalogIndex(games[i], index: i, catalogIndex: &catalogIndex)
        }
        return catalogIndex
    }

    private static func registerInCatalogIndex(
        _ game: CloudGame,
        index: Int,
        catalogIndex: inout CatalogIndex
    ) {
        if !game.id.isEmpty { catalogIndex.byProductId[game.id] = index }
        let conceptKey = conceptPlatformKey(game)
        if !conceptKey.isEmpty { catalogIndex.byConceptId[conceptKey] = index }
        if !game.entitlementId.isEmpty, game.entitlementId != game.id {
            catalogIndex.byProductId[game.entitlementId] = index
        }
    }

    private static func findCatalogIndexForOwned(_ owned: CloudGame, catalogIndex: CatalogIndex) -> Int {
        if !owned.id.isEmpty, let idx = catalogIndex.byProductId[owned.id] { return idx }
        if !owned.entitlementId.isEmpty, let idx = catalogIndex.byProductId[owned.entitlementId] { return idx }
        if !owned.storeProductId.isEmpty, let idx = catalogIndex.byProductId[owned.storeProductId] { return idx }
        // Match by conceptId + platform so an owned PS4 edition does not match a PS5-only catalog
        // entry (and vice-versa); cross-gen editions stay as separate library cards.
        let conceptKey = conceptPlatformKey(owned)
        if !conceptKey.isEmpty, let idx = catalogIndex.byConceptId[conceptKey] { return idx }
        return -1
    }
}

File: android/app/src/main/java/com/metallic/chiaki/cloudplay/api/PsCloudOwnership.kt
// SPDX-License-Identifier: LicenseRef-AGPL-3.0-only-OpenSSL

package com.metallic.chiaki.cloudplay.api

import com.metallic.chiaki.cloudplay.model.CloudGame
import org.json.JSONObject

object PsCloudOwnership
{
	const val PAGE_SIZE = 300
	const val PAGE_COOLDOWN_MS = 100L

	data class Entitlement(
		val id: String,
		val productId: String,
		val activeFlag: Boolean,
		val packageType: String,
		val name: String,
		val conceptId: String,
		val featureType: Int   // PSN feature_type: 3=full game, 1=trial/free, 0=add-on/DLC
	)

	private data class CatalogIndex(
		val byProductId: MutableMap<String, Int>,
		val byConceptId: MutableMap<String, Int>
	)

	fun filterOwnedPs5Games(entitlements: List<Entitlement>): List<Entitlement>
	{
		return entitlements.filter { ent ->
			// Previously required packageType == "PSGD" (PS5 only), which dropped owned PS4
			// titles (e.g. God of War 2018) and PS3 titles. Accept every active game entitlement;
			// streamability is enforced downstream by the cross-reference (deduped by conceptId),
			// so non-streamable / add-on entitlements are harmlessly dropped there.
			ent.activeFlag &&
				!ent.productId.startsWith("IP") &&
				!ent.productId.startsWith("SUB") &&
				// Hide EXTRAS: feature_type==0 is DLC/add-ons/themes/avatars/tracks, never a base game
				// (games are ft 1=trial/free or 3/5=full). Safe -- it can never hide a game.
				ent.featureType != 0
		}
	}

	/** Normalize a conceptId (imagic encodes it as a number) to a non-empty string, else "". */
	private fun conceptIdString(value: Any?): String = when (value)
	{
		is Number -> value.toLong().let { if (it > 0) it.toString() else "" }
		is String -> value
		else -> ""
	}

	fun parseEntitlement(obj: JSONObject): Entitlement?
	{
		val id = obj.optString("id", "")
		if (id.isEmpty()) return null
		val gameMeta = obj.optJSONObject("game_meta") ?: JSONObject()
		val name = gameMeta.optString("name", id)
		val conceptId = conceptIdString(gameMeta.opt("conceptId"))
			.ifEmpty { conceptIdString(gameMeta.opt("concept_id")) }
			.ifEmpty { conceptIdString(obj.opt("conceptId")) }
		return Entitlement(
			id = id,
			productId = obj.optString("product_id", ""),
			activeFlag = obj.optBoolean("active_flag", false),
			packageType = gameMeta.optString("package_type", ""),
			name = name,
			conceptId = conceptId,
			featureType = obj.optInt("feature_type", 0)
		)
	}

	fun crossReferenceOwnedGames(
		filteredEntitlements: List<Entitlement>,
		publicCatalog: List<CloudGame>,
		plusLibrarySupplement: List<CloudGame> = emptyList(),
		productIdAliases: Map<String, String> = emptyMap(),
		componentIdsByProductId: Map<String, List<String>> = emptyMap(),
	): List<CloudGame>
	{
		val catalogMap = catalogMapFirstWins(publicCatalog)
		for ((alias, canonical) in productIdAliases)
		{
			if (alias in catalogMap)
				continue
			catalogMap[canonical]?.let { catalogMap[alias] = it }
		}
		val supplementMap = catalogMapFirstWins(plusLibrarySupplement)
		val browseStableKey = buildStableKeyIndex(publicCatalog)
		val supplementStableKey = buildStableKeyIndex(plusLibrarySupplement)
		val browseByConcept = buildConceptIdIndex(publicCatalog)
		val supplementByConcept = buildConceptIdIndex(plusLibrarySupplement)
		val byKey = linkedMapOf<String, CloudGame>()
		val byKeyRank = mutableMapOf<String, Int>()

		// Enrich one matched catalog row into an owned CloudGame and dedupe it into byKey, keeping OUR
		// convention (conceptId+PLATFORM dedupe, canonical-entitlement rank). Called once for a direct
		// match, or once per component for a bundle (upstream PR #15 bundle-sibling matching).
		fun emit(meta: CloudGame, ent: Entitlement)
		{
			val displayName = meta.name.ifEmpty { ent.name }
			val game = meta.copy(
				name = displayName,
				isOwned = true,
				entitlementId = ent.id,
				storeProductId = ent.productId,
				featureType = ent.featureType
			)
			val key = ownedDedupeKey(meta, ent)
			val candidateRank = ownedStreamRank(ent)
			if (byKey[key] == null)
			{
				byKey[key] = game
				byKeyRank[key] = candidateRank
			}
			// Keep the best streaming candidate: the canonical full-game entitlement (its product_id is
			// the real streamable game, not a DLC/bonus product Gaikai rejects).
			else if (candidateRank > (byKeyRank[key] ?: -1))
			{
				byKey[key] = game
				byKeyRank[key] = candidateRank
			}
		}

		for (ent in filteredEntitlements)
		{
			val stable = productIdStableKey(ent.productId)
			val entStable = productIdStableKey(ent.id)
			val skipStableDemo = ent.name.contains("demo", ignoreCase = true)
			val meta = when {
				ent.productId.isNotEmpty() && catalogMap.containsKey(ent.productId) ->
					catalogMap[ent.productId]
				ent.id.isNotEmpty() && catalogMap.containsKey(ent.id) ->
					catalogMap[ent.id]
				// conceptId is region-stable; product IDs are region-prefixed (EP9000 vs UP9000).
				ent.conceptId.isNotEmpty() && browseByConcept.containsKey(ent.conceptId) ->
					browseByConcept[ent.conceptId]
				ent.conceptId.isNotEmpty() && supplementByConcept.containsKey(ent.conceptId) ->
					supplementByConcept[ent.conceptId]
				ent.productId.isNotEmpty() && ent.id == ent.productId
					&& supplementMap.containsKey(ent.productId) ->
					supplementMap[ent.productId]
				stable != null && !skipStableDemo && browseStableKey.containsKey(stable) ->
					browseStableKey[stable]
				stable != null && !skipStableDemo && supplementStableKey.containsKey(stable) ->
					supplementStableKey[stable]
				// Stable-key match on the ENTITLEMENT id (upstream PR #15): catches cross-gen / upgrade
				// entitlement ids whose stable key matches a catalog row even when product_id did not.
				entStable != null && !skipStableDemo && browseStableKey.containsKey(entStable) ->
					browseStableKey[entStable]
				entStable != null && !skipStableDemo && supplementStableKey.containsKey(entStable) ->
					supplementStableKey[entStable]
				else -> null
			}

			if (meta != null)
			{
				emit(meta, ent)
				continue
			}

			// Bundle-sibling expansion (upstream PR #15): a bundle entitlement (e.g. RE7 Gold) has no
			// direct catalog row, but its component entitlement ids each map to a component game.
			val seenPids = mutableSetOf<String>()
			for (siblingId in componentIdsByProductId[ent.productId] ?: emptyList())
			{
				val siblingMeta = when {
					catalogMap.containsKey(siblingId) -> catalogMap[siblingId]
					supplementMap.containsKey(siblingId) -> supplementMap[siblingId]
					else -> {
						val s2 = productIdStableKey(siblingId)
						if (s2 != null && !skipStableDemo) browseStableKey[s2] ?: supplementStableKey[s2] else null
					}
				} ?: continue
				if (siblingMeta.productId.isEmpty() || seenPids.contains(siblingMeta.productId)) continue
				seenPids.add(siblingMeta.productId)
				emit(siblingMeta, ent)
			}
		}

		return byKey.values.toList()
	}

	// Edition identity = conceptId + PLATFORM (matching the catalog's edition key), so a cross-gen
	// title owned on both PS4 and PS5 stays as two separate library entries instead of collapsing
	// into one. Same-platform duplicate SKUs (a remaster's add-ons) still merge.
	private fun ownedDedupeKey(meta: CloudGame, ent: Entitlement): String
	{
		if (meta.conceptId.isNotEmpty()) return "c:${meta.conceptId}:${platformToken(ent.productId)}"
		if (meta.productId.isNotEmpty()) return "p:${meta.productId}"
		if (ent.id.isNotEmpty()) return "e:${ent.id}"
		return "u:${meta.productId}:${ent.id}"
	}

	/** Platform token from a product id (CUSA = PS4, PPSA = PS5). */
	private fun platformToken(productId: String): String = when
	{
		productId.contains("PPSA") -> "ps5"
		productId.contains("CUSA") -> "ps4"
		else -> ""
	}

	/** A full-game entitlement (vs add-on/avatar): base game has a *GD package_type. */
	private fun isFullGameEntitlement(ent: Entitlement): Boolean =
		ent.featureType == 3 || ent.packageType.endsWith("GD")

	// Rank an owned entitlement as THE streaming candidate for its edition (higher = preferred).
	// Bonus/upgrade SKUs collapse to the same conceptId+platform as the base game; package/feature
	// flags don't disambiguate (Death Stranding DC's "Bonus Content" is also PSGD + feature_type 3).
	// The reliable signal: the base game's entitlement id EQUALS its product_id, while bonus/upgrade
	// SKUs carry a different id -- so prefer the canonical full-game entitlement.
	private fun ownedStreamRank(ent: Entitlement): Int
	{
		var rank = 0
		if (ent.productId.isNotEmpty() && ent.productId == ent.id) rank += 4 // canonical base-game SKU
		if (isFullGameEntitlement(ent)) rank += 2
		if (ent.id.isNotEmpty()) rank += 1
		return rank
	}

	/** conceptId + platform; the owned product id (storeProductId) takes precedence so the owned
	 * edition's platform is used, else the catalog product id. */
	private fun conceptPlatformKey(game: CloudGame): String
	{
		if (game.conceptId.isEmpty()) return ""
		val pid = if (game.storeProductId.isNotEmpty()) game.storeProductId else game.productId
		return "${game.conceptId}|${platformToken(pid)}"
	}

	private fun catalogMapFirstWins(games: List<CloudGame>): MutableMap<String, CloudGame>
	{
		val map = linkedMapOf<String, CloudGame>()
		for (game in games)
		{
			if (game.productId.isNotEmpty() && game.productId !in map)
				map[game.productId] = game
		}
		return map
	}

	/** Tokenize on '-' and '_'; identity is all tokens except the last (store SKU). */
	private fun productIdStableKey(productId: String): String?
	{
		if (productId.isEmpty())
			return null
		val tokens = mutableListOf<String>()
		for (dashPart in productId.split('-'))
		{
			for (token in dashPart.split('_'))
			{
				if (token.isNotEmpty())
					tokens.add(token)
			}
		}
		if (tokens.size < 2)
			return null
		return tokens.dropLast(1).joinToString("|")
	}

	private fun buildStableKeyIndex(games: List<CloudGame>): Map<String, CloudGame>
	{
		val index = linkedMapOf<String, CloudGame>()
		for (game in games)
		{
			val key = productIdStableKey(game.productId) ?: continue
			if (key !in index)
				index[key] = game
		}
		return index
	}

	private fun buildConceptIdIndex(games: List<CloudGame>): Map<String, CloudGame>
	{
		val index = linkedMapOf<String, CloudGame>()
		for (game in games)
		{
			if (game.conceptId.isNotEmpty() && game.conceptId !in index)
				index[game.conceptId] = game
		}
		return index
	}

	fun mergeOwnedIntoBrowseCatalog(
		browseCatalog: List<CloudGame>,
		ownedCrossRef: List<CloudGame>,
		addUnmatched: Boolean = true   // false = only mark ownership (Catalog tab), never add
	): List<CloudGame>
	{
		val games = browseCatalog.toMutableList()
		val catalogIndex = buildCatalogIndex(games)

		for (owned in ownedCrossRef)
		{
			// Trials / free-to-play (feature_type 1) are kept as their OWN card so the user can Stream
			// the trial/free build, while the full version still shows separately as a not-owned
			// "Add Game" card -- so a trial must NOT collapse into the full-game catalog entry.
			val catalogMatch = if (owned.featureType == 1) -1 else findCatalogIndexForOwned(owned, catalogIndex)
			if (catalogMatch >= 0)
			{
				val existing = games[catalogMatch]
				games[catalogMatch] = existing.copy(
					isOwned = true,
					entitlementId = owned.entitlementId.ifEmpty { existing.entitlementId },
					storeProductId = owned.storeProductId.ifEmpty { existing.storeProductId }
				)
				continue
			}

			if (!addUnmatched) continue
			val entry = owned.copy(isOwned = true)
			registerInCatalogIndex(entry, games.size, catalogIndex)
			games.add(entry)
		}

		return games.sortedWith(
			compareByDescending<CloudGame> { it.isOwned }
				.thenBy { it.name.lowercase() }
		)
	}

	fun streamingIdentifier(game: CloudGame): String
	{
		if (game.serviceType.equals("pscloud", ignoreCase = true))
		{
			// Stream the owned PRODUCT id (storeProductId) before the entitlement id: for cross-gen
			// upgrades the entitlement id is the stale original SKU Gaikai has no game for.
			if (game.storeProductId.isNotEmpty()) return game.storeProductId
			if (game.entitlementId.isNotEmpty()) return game.entitlementId
		}
		return game.productId
	}

	// A PlayStation title id encodes its platform: CUSAxxxxx = PS4, PPSAxxxxx = PS5. This is more
	// reliable than the catalog device list and decides the streaming path: PS4 catalog titles go
	// through Kamaji (psnow) to acquire the streaming entitlement, PS5 streams directly (pscloud).
	fun streamPlatform(game: CloudGame): String
	{
		// Prefer the OWNED product id (storeProductId): for a cross-gen title you upgraded, the catalog
		// productId may be the other generation (Alan Wake catalog = PS4 CUSA, but you own the PS5 PPSA).
		val p = game.storeProductId.ifEmpty { game.productId.ifEmpty { game.entitlementId } }
		return when
		{
			p.contains("PPSA") -> "ps5"
			p.contains("CUSA") -> "ps4"
			else -> game.platform.ifEmpty { "ps5" }
		}
	}

	/** Real legacy PS Now games stay psnow; otherwise route by title-id platform. */
	fun streamServiceType(game: CloudGame): String
	{
		if (game.serviceType.equals("psnow", ignoreCase = true)) return "psnow"
		return if (streamPlatform(game) == "ps4") "psnow" else "pscloud"
	}

	/** Identifier for startCompleteCloudSession: psnow sends the product id (Kamaji converts it
	 *  and acquires via PS Plus); pscloud sends the owned entitlement id (direct). */
	fun streamIdentifier(game: CloudGame): String
	{
		return if (streamServiceType(game) == "psnow") game.productId.ifEmpty { streamingIdentifier(game) }
		else streamingIdentifier(game)
	}

	private fun buildCatalogIndex(games: List<CloudGame>): CatalogIndex
	{
		val byProductId = mutableMapOf<String, Int>()
		val byConceptId = mutableMapOf<String, Int>()
		for (i in games.indices)
			registerInCatalogIndex(games[i], i, CatalogIndex(byProductId, byConceptId))
		return CatalogIndex(byProductId, byConceptId)
	}

	private fun registerInCatalogIndex(game: CloudGame, index: Int, catalogIndex: CatalogIndex)
	{
		if (game.productId.isNotEmpty())
			catalogIndex.byProductId[game.productId] = index
		val conceptKey = conceptPlatformKey(game)
		if (conceptKey.isNotEmpty())
			catalogIndex.byConceptId[conceptKey] = index
		if (game.entitlementId.isNotEmpty() && game.entitlementId != game.productId)
			catalogIndex.byProductId[game.entitlementId] = index
	}

	private fun findCatalogIndexForOwned(owned: CloudGame, catalogIndex: CatalogIndex): Int
	{
		if (owned.productId.isNotEmpty() && catalogIndex.byProductId.containsKey(owned.productId))
			return catalogIndex.byProductId.getValue(owned.productId)
		if (owned.entitlementId.isNotEmpty() && catalogIndex.byProductId.containsKey(owned.entitlementId))
			return catalogIndex.byProductId.getValue(owned.entitlementId)
		if (owned.storeProductId.isNotEmpty() && catalogIndex.byProductId.containsKey(owned.storeProductId))
			return catalogIndex.byProductId.getValue(owned.storeProductId)
		// Match by conceptId + platform so an owned PS4 edition does not match a PS5-only catalog
		// entry (and vice-versa); cross-gen editions stay as separate library cards.
		val conceptKey = conceptPlatformKey(owned)
		if (conceptKey.isNotEmpty() && catalogIndex.byConceptId.containsKey(conceptKey))
			return catalogIndex.byConceptId.getValue(conceptKey)
		return -1
	}
}
